lib/ex_plasma/transaction/type/payment_v1.ex

```diff
@@ -1,150 +1,249 @@
-  ...>  [
-  ...>    [
-  ...>      <<1>>, 
-  ...>      [<<29, 246, 47, 41, 27, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>, <<46, 38, 45, 41, 28, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>, <<1>>]
-  ...>    ]
-  ...>  ], 
-  ...>  0,
-  ...>  <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>
-  ...>]
-  iex> ExPlasma.Transaction.Type.PaymentV1.to_map(rlp)
-  %ExPlasma.Transaction{
-  	inputs: [
-  		%ExPlasma.Output{
-  			output_data: nil,
-  			output_id: %{blknum: 0, oindex: 0, position: 0, txindex: 0},
-  			output_type: nil
-  		}
-  	],
-  	metadata: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
-  	outputs: [
-  		%ExPlasma.Output{
-  			output_data: %{
-  				amount: 1,
-  				output_guard: <<29, 246, 47, 41, 27, 46, 150, 159, 176, 132, 157, 153,
-  					217, 206, 65, 226, 241, 55, 0, 110>>,
-  				token: <<46, 38, 45, 41, 28, 46, 150, 159, 176, 132, 157, 153, 217, 206,
-  					65, 226, 241, 55, 0, 110>>
-  			},
-  			output_id: nil,
-  			output_type: 1
-  		}
-  	],
-  	sigs: [],
-  	tx_data: 0,
-  	tx_type: 1
-  }
+  Only validates that the RLP is structurally correct.
+  Does not perform any other kind of validation, use validate/1 for that.
   """
-  @impl Transaction
-  @spec to_map(list()) :: Transaction.t()
-  def to_map(rlp) when is_list(rlp), do: do_to_map(rlp)
-
-  defp do_to_map([sigs, tx_type, inputs, outputs, "", metadata]),
-    do: do_to_map([sigs, tx_type, inputs, outputs, 0, metadata])
-
-  defp do_to_map([sigs, <<tx_type>>, inputs, outputs, tx_data, metadata]),
-    do: do_to_map([sigs, tx_type, inputs, outputs, tx_data, metadata])
-
-  defp do_to_map([sigs, tx_type, inputs, outputs, tx_data, metadata]) do
-    %ExPlasma.Transaction{
-      sigs: sigs,
-      tx_type: tx_type,
-      inputs: Enum.map(inputs, &ExPlasma.Output.decode_id/1),
-      outputs: Enum.map(outputs, &ExPlasma.Output.decode/1),
-      tx_data: tx_data,
-      metadata: metadata
-    }
+  @spec to_map(PaymentV1.t(), list()) :: {:ok, PaymentV1.t()} | {:error, mapping_error()}
+  def to_map(%PaymentV1{}, [<<@tx_type>>, inputs_rlp, outputs_rlp, tx_data_rlp, metadata_rlp]) do
+    with {:ok, inputs} <- decode_inputs(inputs_rlp),
+         {:ok, outputs} <- decode_outputs(outputs_rlp),
+         {:ok, tx_data} <- decode_tx_data(tx_data_rlp),
+         {:ok, metadata} <- decode_metadata(metadata_rlp) do
+      {:ok,
+       %PaymentV1{
+         tx_type: @tx_type,
+         inputs: inputs,
+         outputs: outputs,
+         tx_data: tx_data,
+         metadata: metadata
+       }}
+    end
   end
 
+  def to_map(_, _), do: {:error, :malformed_transaction}
+
+  defp decode_inputs(inputs_rlp) do
+    {:ok, Enum.map(inputs_rlp, &Output.decode_id(&1))}
+  rescue
+    _ -> {:error, :malformed_inputs}
+  end
+
+  defp decode_outputs(outputs_rlp) do
+    {:ok, Enum.map(outputs_rlp, &Output.decode(&1))}
+  rescue
+    _ -> {:error, :malformed_outputs}
+  end
```

Comment:
this method of `rescuing` everything was copied over from elixir-omg I presume? Is there a better way?

Reply:
I guess there would be if I update the output decoding to handle errors properly but there is already a lot of refactor in this PR so I wanted to avoid touching outputs for now.

Reply:
for now it should be enough to catch what kind of error you're expecting if decoding fails?

Reply:
Problem is that it'll crash for now, hence the rescue, do you mean I should add explicit clauses to the rescue instead of `_ ->`?

Reply:
yeah! is it feasable?

Reply:
Hmm, not sure if it really adds any value until we improve error handling in `Output.decode`. This are the errors raised if I log them when running the tests:
```
%Protocol.UndefinedError{description: "", protocol: Enumerable, value: "*"}
%FunctionClauseError{
  args: nil,
  arity: 1,
  clauses: nil,
  function: :do_decode,
  kind: nil,
  module: ExPlasma.Output
}
%FunctionClauseError{
  args: nil,
  arity: 1,
  clauses: nil,
  function: :to_map,
  kind: nil,
  module: ExPlasma.Output.Type.GenericPayment
}
%ArgumentError{message: "argument error"}
%ArgumentError{message: "output type 232 does not exist."}
%ArgumentError{message: "argument error"}
%Protocol.UndefinedError{description: "", protocol: Enumerable, value: 123}
%FunctionClauseError{
  args: nil,
  arity: 2,
  clauses: nil,
  function: :decode,
  kind: nil,
  module: ExRLP.Decode
}
```